lambeq/training/nelder_mead_optimizer.py

```diff
@@ -0,0 +1,427 @@
+                self.sim.ndim != 2
+                or self.sim.shape[0] != self.sim.shape[1] + 1
+            ):
+                raise ValueError(
+                    '`initial_simplex` should be an array of shape'
+                    f'({N+1},{N})'
+                )
+            if len(self.model.weights) != self.sim.shape[1]:
+                raise ValueError(
+                    'Size of `initial_simplex` is not consistent with'
+                    '`model.weights`'
+                )
+            N = self.sim.shape[1]
+
+        self.sim = self.project(self.sim)
+        self.one2np1 = list(range(1, N + 1))
+        self.fsim = np.full((N + 1,), np.inf, dtype=float)
+        self.N = N
+        self.first_iter = True  # flag for first iteration
+
+    def backward(self, batch: tuple[Iterable[Any], np.ndarray]) -> float:
+        """Calculate the gradients of the loss function.
+
+        The gradients are calculated with respect to the model
+        parameters.
+
+        Parameters
+        ----------
+        batch : tuple of Iterable and numpy.ndarray
+            Current batch. Contains an Iterable of diagrams in index 0,
+            and the targets in index 1.
+
+        Returns
+        -------
+        float
+            The calculated loss.
+
+        """
+
+        diagrams, targets = batch
+        diags_gen = flatten(diagrams)
+
+        # the symbolic parameters
+        parameters = self.model.symbols
+
+        relevant_params = set.union(
+            *[diag.free_symbols for diag in diags_gen]
+        )
+        mask = np.array(
+            [1 if sym in relevant_params else 0 for sym in parameters]
+        )
+
+        # Initialize ind, sim and fsim
+        if self.first_iter:
+            try:
+                for k in range(self.N + 1):
+                    self.fsim[k] = self.objective_func(diagrams,
+                                                       targets,
+                                                       self.sim[k])
+            except RuntimeError:
+                pass
+            finally:
+                self.ind = np.argsort(self.fsim)
+                self.sim = np.take(self.sim, self.ind, 0)
+                self.fsim = np.take(self.fsim, self.ind, 0)
+                self.first_iter = False  # set flag to false
+
+        try:
+            if (
+                np.max(np.ravel(np.abs(self.sim[1:] - self.sim[0])))
+                <= self.xatol
+                and np.max(np.abs(self.fsim[0] - self.fsim[1:])) <= self.fatol
+            ):
+                raise RuntimeError(
+                    'xatol and fatol termination conditions are satisfied.'
+                )
+
+            xbar = np.add.reduce(self.sim[:-1], 0) / self.N
+            xr = (1 + self.rho) * xbar - self.rho * self.sim[-1]
+            xr = self.project(xr)
+            fxr = self.objective_func(diagrams, targets, xr)
+            shrink = False
+
+            if fxr < self.fsim[0]:
+                xe = (
```

Comment:
Could this be better as `xe = xbar + self.rho * self.chi * (xbar + self.sim[-1])`? (similarly for a later calculation for `xc`)